core/wait/wait_test.go

```diff
@@ -169,25 +267,239 @@ func TestSetTimeout(t *testing.T) {
+			handlerRetryLimitTempErr:       5,
+			contextTimeout:                 1000 * time.Millisecond,
+			wantCheckFnNumberCalls:         1,
+			wantErr:                        true,
+		},
+		{
+			desc:                           "retry_limit_temp_err_2",
+			checkFnNumberCallsToFinishWait: 999999,
+			checkFnWaitSucceeds:            true,
+			checkFnNumberCallsUntilErr:     1,
+			checkFnReturnsTempErr:          true,
+			handlerSleepBeforeWait:         0,
+			handlerThrottle:                30 * time.Millisecond,
+			handlerTimeout:                 1000 * time.Millisecond,
+			handlerRetryLimitTempErr:       5,
+			contextTimeout:                 1000 * time.Millisecond,
+			wantCheckFnNumberCalls:         5,
+			wantErr:                        true,
+		},
+		{
+			desc:                           "retry_limit_temp_err_3",
+			checkFnNumberCallsToFinishWait: 3,
+			checkFnWaitSucceeds:            true,
+			checkFnNumberCallsUntilErr:     1,
+			checkFnReturnsTempErr:          true,
+			handlerSleepBeforeWait:         0,
+			handlerThrottle:                30 * time.Millisecond,
+			handlerTimeout:                 1000 * time.Millisecond,
+			handlerRetryLimitTempErr:       5,
+			contextTimeout:                 1000 * time.Millisecond,
+			wantCheckFnNumberCalls:         3,
+			wantErr:                        false,
+		},
+	} {
+		t.Run(tt.desc, func(t *testing.T) {
+			type respType struct{}
+
+			numberCheckFnCalls := 0
+			checkFn := func() (waitFinished bool, response *respType, err error) {
+				numberCheckFnCalls++
+				if numberCheckFnCalls == tt.checkFnNumberCallsToFinishWait {
+					if tt.checkFnWaitSucceeds {
+						return true, &respType{}, nil
+					}
+					return true, &respType{}, fmt.Errorf("the async action couldn't be done")
+				}
+
+				if numberCheckFnCalls < tt.checkFnNumberCallsUntilErr {
+					return false, nil, nil
+				}
+
+				if tt.checkFnReturnsTempErr {
+					return false, nil, &oapiError.GenericOpenAPIError{
+						StatusCode:   RetryHttpErrorStatusCodes[0],
+						ErrorMessage: "something bad happened when checking if the async action was finished",
+					}
+				}
+				return false, nil, fmt.Errorf("something bad happened when checking if the async action was finished")
+			}
+			handler := AsyncActionHandler[respType]{
+				checkFn:           checkFn,
+				sleepBeforeWait:   tt.handlerSleepBeforeWait,
+				throttle:          tt.handlerThrottle,
+				timeout:           tt.handlerTimeout,
+				retryLimitTempErr: tt.handlerRetryLimitTempErr,
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), tt.contextTimeout)
+			defer cancel()
+
+			resp, err := handler.WaitWithContext(ctx)
+
+			if tt.wantErr && (err == nil) {
+				t.Errorf("expected error but got none")
+			}
+			if !tt.wantErr && (err != nil) {
+				t.Errorf("expected no error but got \"%v\"", err)
+			}
+			if (err == nil) && (resp == nil) {
+				t.Errorf("got nil err but nil resp")
+			}
```

Comment:
Similar to a comment I made above, can't we have `checkFn`s returning nil? If the API doesn't have a response body

Reply:
Check comment above